dotnet/src/Connectors/Connectors.Anthropic/Core/Models/Message/TypeJsonDyscriminator.cs

```diff
@@ -0,0 +1,121 @@
+
+// Temporary solution from https://github.com/dotnet/runtime/issues/72604
+// TODO: Remove this once we move to .NET 9
+
+internal static class JsonTypeDiscriminatorHelper
+{
+    internal static IJsonTypeInfoResolver TypeInfoResolver { get; } = new DefaultJsonTypeInfoResolver
+    {
+        Modifiers =
+        {
+            static typeInfo =>
+            {
+                var propertyNamingPolicy = typeInfo.Options.PropertyNamingPolicy;
+
+                // Temporary hack to ensure subclasses of abstract classes will always include the type field
+                if (typeInfo.Type.BaseType is { IsAbstract: true } &&
+                    typeInfo.Type.BaseType.GetCustomAttributes<HackyJsonDerivedAttribute>().Any())
+                {
+                    var discriminatorPropertyName = propertyNamingPolicy?.ConvertName("type") ?? "type";
+                    if (typeInfo.Properties.All(p => p.Name != discriminatorPropertyName))
+                    {
+                        var discriminatorValue = typeInfo.Type.BaseType
+                            .GetCustomAttributes<HackyJsonDerivedAttribute>()
+                            .First(attr => attr.Subtype == typeInfo.Type).TypeDiscriminator;
+                        var propInfo = typeInfo.CreateJsonPropertyInfo(typeof(string), discriminatorPropertyName);
+                        propInfo.Get = _ => discriminatorValue;
+                        typeInfo.Properties.Insert(0, propInfo);
+                    }
+                }
+            },
+        },
+    };
+}
+
+/// <summary>
+/// Same as <see cref="JsonDerivedTypeAttribute"/> but used for the hack below. Necessary because using the built-in
+/// attribute will lead to NotSupportedExceptions.
+/// </summary>
+[AttributeUsage(AttributeTargets.Class | AttributeTargets.Interface, AllowMultiple = true, Inherited = false)]
+internal sealed class HackyJsonDerivedAttribute(Type subtype, string typeDiscriminator) : Attribute
+{
+    public Type Subtype { get; internal set; } = subtype;
+    public string TypeDiscriminator { get; internal set; } = typeDiscriminator;
+}
+
+internal sealed class PolymorphicJsonConverterFactory : JsonConverterFactory
+{
+    public override bool CanConvert(Type typeToConvert)
+    {
+        return typeToConvert.IsAbstract && typeToConvert.GetCustomAttributes<HackyJsonDerivedAttribute>().Any();
+    }
+
+    public override JsonConverter? CreateConverter(Type typeToConvert, JsonSerializerOptions options)
+    {
+        return (JsonConverter?)Activator.CreateInstance(
+            typeof(PolymorphicJsonConverter<>).MakeGenericType(typeToConvert), options);
+    }
+}
+
+/// <summary>
+/// A temporary hack to support deserializing JSON payloads that use polymorphism but don't specify type as the first field.
+/// Modified from https://github.com/dotnet/runtime/issues/72604#issuecomment-1440708052.
+/// </summary>
+internal sealed class PolymorphicJsonConverter<T> : JsonConverter<T>
```

Comment:
Move to another file `PolymorphicJsonConverterOfT.cs`

```suggestion
internal sealed class PolymorphicJsonConverter<T> : JsonConverter<T>
```

Reply:
I don't like this class and all of the contents in this file. Maybe you have a better idea how to fix the problem when the metadata type is not the first field in json?

Reply:
Im not fully into the details of the need of those classes for Json conversions, are all of those needed? What is the main impact of having almost none of them?

Reply:
The problem is System.Text.Json requires to metadata fields be on first position in json, but Claude returns Type metadata in not correct order.